### ISSUE TITLE
Is there any demand for supporting other instruments or kits?

### ISSUE THREAD

**Issue by jwfoley** [jwfoley/bioanalyzeR#14]
Currently this package supports only the Agilent Bioanalyzer and TapeStation because those are the two instruments I use myself, and only their DNA and RNA kits because that allowed me to take a few small shortcuts. However, I reuse a lot of code between the two platforms because many of the problems are the same, so it might not be too much trouble to support data from other similar instruments, from the Agilent Fragment Analyzer to even handmade gels analyzed in ImageJ. The problem is stress-testing the package with imperfect real-world data instead of the demo samples provided by the manufacturer.

So please respond here if you are interested in using this package on another kind of data and you are willing to help test it.

**Comment by rmvpaeme** [comment]
Hi, our research group often uses fragment analyser and Femto PULSE data. We have written a custom script for visualising the electropherograms in [this manuscript](https://www.biorxiv.org/content/biorxiv/early/2020/04/24/2020.04.23.055293/F2.large.jpg?width=800&height=600&carousel=1). However, these scripts don't scale well and we would prefer to use R to generate these figures instead of Prosize (the software provided by Agilent). We were looking for an R package that does the job and found this page :-). We have sufficient test data, so if you want to work together for this let us know.

**Comment by jwfoley** [comment]
Thanks for your interest. I've never used those instruments but I took a look at the software (it appears to be the same for both as well as ZAG?), and at least with the provided demo files it appears you can export the raw data and peak table into plain CSV files, which would be very easy to import. For detailed metadata you would still need the original data files as well, but it looks like it might be safe to skip those for convenience. Unfortunately the original raw data are in a less easily parseable binary format (`.raw` file) - does your script read that or just the exported CSVs?

When I find some time I'll write some functions to read these files into the `electrophoresis` data structure, and then all the existing analysis tools should probably work the same since they're already platform-independent between the Bioanalyzer and TapeStation (which use very different physics). Would you be willing to test a development pre-release on your data? Real-world data contain artifacts and errors that don't appear in the perfect demo data so it's hard to account for all the corner cases without some trial and error.

**Comment by rmvpaeme** [comment]
Yes, I used the export to csv function in the Prosize software. 

Sure, we can test the development versions. There is already some [example data here](https://github.com/rmvpaeme/cfRRBS_tube_study/tree/master/data/femto_data) and [this](https://github.com/rmvpaeme/cfRRBS_tube_study/blob/master/Markdowns/cfRRBS_tubestudy_DataAnalysis.Rmd#L147) is the code from the figure I linked earlier.

**Comment by rmvpaeme** [comment]
My colleague @jpvw looked further into this, and there is a way easier export function in Prosize which gives *electropherograms.csv files with the sizing included (in contrast to the files above which give time instead of size):

[2020 11 30 13H 23M Electropherogram.txt](https://github.com/jwfoley/bioanalyzeR/files/5695303/2020.11.30.13H.23M.Electropherogram.txt)

```
library(tidyverse)
df <- read_csv("2020 11 30 13H 23M Electropherogram.csv")
ggplot(df %>% gather(sample_name, value, -(`Size (bp)`)), aes(x = `Size (bp)`, y = value, col = sample_name)) +
     geom_line() + facet_wrap(~sample_name, scales = "free_y") + theme_bw() + lims(x = c(0,1000))
```

![ggprosize](https://user-images.githubusercontent.com/45523344/102206922-98572d80-3ecd-11eb-9246-c49dc26d24d2.png)

**Comment by jwfoley** [comment]
I think I have a crude version working, though I haven't tested it very thoroughly yet. Please clone from the `prosize` branch or do `install.packages` on the attached file. For now you have to export the data with a specific configuration, which you can see with `vignette("bioanalyzeR")` or [here](https://stanford.edu/~jwfoley/bioanalyzeR-prosize.html#prosize-fragment-analyzer-zag-dna-analyzer-femto-pulse). I haven't updated the rest of the documentation yet.

Use `read.electrophoresis` to open an electropherogram CSV file and then check `qc.electrophoresis` to make sure my math is right. Please test with a wide range of samples and let me know if you find any problems. In particular, see what happens with anomalous data (instrument errors, failed samples, etc.) since nothing like that is provided with the software.

[bioanalyzeR-prosize.tar.gz](https://github.com/jwfoley/bioanalyzeR/files/5706466/bioanalyzeR-prosize.tar.gz)

**Comment by jwfoley** [comment]
One other question: how do ladders work on these instruments? Are you always required to load the ladder in the final well of a row/plate, or can you put it wherever you want, or can you use a different ladder in each row?

**Comment by rmvpaeme** [comment]
There are three options as far as I know:
1) The ladder is loaded in well 12. I think this is the default setting. 
2) The ladder is loaded in a different well, which then has to be specified in the prosize software.
3) No ladder is loaded, but a ladder that was run on a previous run can be loaded (or a ladder from the example data).

Thanks for your very quick work, we'll check it out and let you know.

**Comment by jwfoley** [comment]
Unfortunately I don't see anything in the exported metadata that indicates which well was designated as the ladder, and there isn't even a consistent naming scheme, so the package just looks for samples with "ladder" in the name and uses the one with the highest well number, in hopes that the user hasn't done anything weird.

I coded around some more corner cases until all the demo runs were compatible, so you can pull the new commits from the `prosize` branch or install this updated tarball: [bioanalyzeR-prosize2.tar.gz](https://github.com/jwfoley/bioanalyzeR/files/5717643/bioanalyzeR-prosize2.tar.gz). I also added an option to read a ZIP archive containing all the CSV files from a batch, which you can see in the updated exporting instructions.

Have you found any problems yet?

**Comment by jwfoley** [comment]
Things seem stable on at least the demo data so I've merged the new code and published it as [release 0.6.0](https://github.com/jwfoley/bioanalyzeR/releases/tag/v0.6.0). Please install that the usual way and open a new issue if you find any problems.
